qupath-gui-fx/src/main/java/qupath/lib/gui/charts/ScatterPlotChart.java

```diff
@@ -0,0 +1,337 @@
+     *
+     * @param name the name of the data series (useful if multiple series will be plotted, otherwise may be null)
+     * @param x    x-values
+     * @param y    y-values
+     * @return a series of data
+     */
+    private static Series<Number, Number> series(String name, double[] x, double[] y) {
+        return series(name, x, y, (List<?>)null);
+    }
+
+    /**
+     * Create a series of data using collections of numeric values, with an associated custom object.
+     *
+     * @param <T>   The type of custom object.
+     * @param name  the name of the data series (useful if multiple series will be plotted, otherwise may be null)
+     * @param x     x-values
+     * @param y     y-values
+     * @param extra array of values to associate with each data point; should be the same length as x and y
+     * @return a series of data
+     */
+    private static <T> Series<Number, Number> series(String name, double[] x, double[] y, T[] extra) {
+        return series(name, x, y, extra == null ? null : Arrays.asList(extra));
+    }
+
+    /**
+     * Create a scatterplot using collections of numeric values, with an associated custom object.
+     *
+     * @param <T>   The type of custom object.
+     * @param name  the name of the data series (useful if multiple series will be plotted, otherwise may be null)
+     * @param x     x-values
+     * @param y     y-values
+     * @param extra list of values to associate with each data point; should be the same length as x and y
+     * @return a series of data
+     */
+    private static <T> Series<Number, Number> series(String name, double[] x, double[] y, List<T> extra) {
+        List<Data<Number, Number>> data = new ArrayList<>();
+        for (int i = 0; i < x.length; i++) {
+            if (extra != null && i < extra.size())
+                data.add(new Data<>(x[i], y[i], extra.get(i)));
+            else
+                data.add(new Data<>(x[i], y[i]));
+        }
+        return series(name, data);
+    }
+
+    /**
+     * Create a series of data from existing data sets.
+     * @param name the name of the data series (useful if multiple series will be plotted, otherwise may be null)
+     * @param data the data points to plot
+     * @return a series of data
+     */
+    private static Series<Number, Number> series(String name, Collection<Data<Number, Number>> data) {
+        return new XYChart.Series<>(name, FXCollections.observableArrayList(data));
+    }
+
+    /**
+     *
+     * Try to select an object if possible (e.g. because a user clicked on it).
+     * @param pathObject the object to select
+     * @param addToSelection if true, add to an existing selection; if false, reset any current selection
+     * @param centerObject if true, try to center it in a viewer (if possible)
+     */
+    private static void tryToSelect(PathObject pathObject, ImageData<?> imageData, QuPathViewer viewer, boolean addToSelection, boolean centerObject) {
```

Comment:
Isn't this code duplication from `Charts`? Is it necessary?

Reply:
Short answer: no
Long answer: I found it difficult to find a good point to refactor the existing code in Charts, as it performs subsampling in a spot that I didn't expect.* But that's not a good reason to not refactor, as the charts class would also benefit.

*possibly resulting in confusing code and/or unexpected outputs

Reply:
I have made a first pass at a refactor, although I don't think I did a good job.

This means that now the following two calls will produce plots with the same number of points, where they didn't before:

```groovy
import qupath.lib.gui.charts.Charts.ScatterChartBuilder

new ScatterChartBuilder()
    .measurements(getDetectionObjects(), "Nucleus: Area", "Nucleus: Perimeter")
    .limitDatapoints(10)
    .show()
    
    
new ScatterChartBuilder()
    .limitDatapoints(10)
    .measurements(getDetectionObjects(), "Nucleus: Area", "Nucleus: Perimeter")
    .show()
```

Reply:
Don't think changing behavior is a big problem, `Charts` was just there for scripting and not very well known. Should mention in the changelog though.

I'm not entirely sure if subsampling should be at the series or chart level. Subsampling all datapoints makes sense, although for rare classes I imagine we could often lose all the instances so they aren't represented at all.

Reply:
I think subsampling the series is the easiest option that is unlikely to have consequences; I'll update the docs to make that clear

Reply:
Yeah, although if there are 10x of one class than another, they might end up having the same number. Which could be ok, I guess... or if we have a limit of 10,000 points but 100 classes then we still end up with a whole lot of points. No option seems without problems somewhere.

Reply:
I guess the nicest way to do it would be to proportionally subsample from all of the series, but I doubt many people will notice which way we do it